public/app/features/dashboard-scene/utils/DashboardModelCompatibilityWrapper.ts

```diff
@@ -232,6 +247,50 @@ class PanelCompatibilityWrapper {
     return [];
   }
 
+  public get targets() {
+    if (this._vizPanel.state.$data instanceof SceneQueryRunner) {
+      return this._vizPanel.state.$data.state.queries;
+    }
+
+    if (this._vizPanel.state.$data instanceof ShareQueryDataProvider) {
+      return [
+        {
+          datasource: {
+            uid: SHARED_DASHBOARD_QUERY,
+            type: 'datasource',
+          },
+          ...this._vizPanel.state.$data.state.query,
+        },
+      ];
+    }
+
+    if (this._vizPanel.state.$data instanceof SceneDataTransformer) {
+      return (this._vizPanel.state.$data.state.$data as SceneQueryRunner).state.queries;
+    }
+
+    return [];
+  }
+
+  public get datasource(): DataSourceRef | null {
+    if (this._vizPanel.state.$data instanceof SceneQueryRunner) {
+      return this._vizPanel.state.$data.state.datasource ?? null;
+    }
+
+    if (this._vizPanel.state.$data instanceof ShareQueryDataProvider) {
+      return { uid: SHARED_DASHBOARD_QUERY, type: 'datasource' };
+    }
+
+    if (this._vizPanel.state.$data instanceof SceneDataTransformer) {
+      if (this._vizPanel.state.$data.state.$data instanceof ShareQueryDataProvider) {
+        return { uid: SHARED_DASHBOARD_QUERY, type: 'datasource' };
+      }
+
+      return (this._vizPanel.state.$data.state.$data as SceneQueryRunner).state.datasource ?? null;
+    }
```

Comment:
Maybe - but for now I didn't see a need for that. I'll refactor this when we identify that the logic for resolving `DataSourceRef` is needed in some other place.

Reply:
@dprokop but is that not true now? this could use getQueryRunnerFor (which is used in a lot of places) if we added the ShareQueryDataProvider check to it. 

this would then simply become  

```ts
public get datasource(): DataSourceRef | null {
  return getQueryRunnerFor(this)?.state.datasource ?? null;
}
```

Reply:
hm, with a close look it's not the same as this does not really want the inner/source query runner just the configured data source.

Reply:
yeah, precisely, hence this is a custom implementation in the compat wrapper.